Add twist effect in scan angle calculation for Cris and code restructure

emilyhcliu opened the pull request:
This PR does two things:
1. Add twist effect in scan angle calculation for CrIS
2. Fix bug in scan angle calculation for IASI
3. Restructure the code to use less if statments

The code handles three cases:
default - need scan position, starting angle and angle increment (for most of the sensors)
IASI - need to restructure scan position
CrIS - need to add twist effect 


This PR fixes Issue #68 and Issue #26

**CrIS Scan Angles**
![cris-fsr_npp_channel_440](https://github.com/user-attachments/assets/7ce370de-3d89-4804-baa9-2686fb7b200e)

**IASI Scan Angles**
![iasi_metop-b_channel_921](https://github.com/user-attachments/assets/6586f7ba-69d8-4a12-bd07-a4d0671b7910)

rmclaren committed:
added oobsbuilder python object

rmclaren committed:
tweaks

rmclaren committed:
tweaks

rmclaren committed:
removed the need to address variables as variables/something..

rmclaren committed:
enhanced logger

rmclaren committed:
more ObsBuilder enhancements

rmclaren committed:
updated uml

rmclaren committed:
enhancec obsbuilder so it can take multiple input files and mappings.

rmclaren committed:
enhancec obsbuilder so it can take multiple input files and mappings.

rmclaren committed:
added apply_mask method to the DataContainer

rmclaren committed:
added wrap transform

rmclaren committed:
added DataContainer::remove method

rmclaren committed:
log string bug fix

rmclaren committed:
ObsBuilder no longer expects the input file in its constructor

rmclaren committed:
ObsBuilder no longer expects the input file in its constructor

rmclaren committed:
ObsBuilder no longer expects the input file in its constructor

rmclaren committed:
fixed bugs with description remove, container mask, and updated err mesg

rmclaren committed:
obs builder updates

rmclaren committed:
obs builder updates, two versions of obs_group

rmclaren committed:
obs builder updates, added more fine grained create obsgroup functions.

rmclaren committed:
no longer running make_description directly

rmclaren committed:
added some obs builder docs

rmclaren committed:
removed mapping files as inputs, fixed some bugs

rmclaren committed:
added ability to add config dict

rmclaren committed:
added ability to add config dict

rmclaren committed:
fixed issue in logger

rmclaren committed:
updated zar encoder to support older versions of python

rmclaren committed:
fixed bug

emilyhcliu committed:
Merge branch 'develop' into feature/obs_builder

rmclaren committed:
removed unused classes

rmclaren committed:
Merge branch 'feature/obs_builder' of https://github.com/NOAA-EMC/bufr-query into feature/obs_builder

rmclaren committed:
fixed bug where we were returning the dictionary key instead of value in create_obs_group_no_cache

rmclaren committed:
fixed apply container mask for string fields

rmclaren committed:
Merge branch 'feature/obs_builder' of https://github.com/NOAA-EMC/bufr-query into feature/obs_builder

rmclaren committed:
added support for string vector support globals

rmclaren committed:
Merge branch 'feature/obs_builder' of https://github.com/NOAA-EMC/bufr-query into feature/obs_builder

rmclaren committed:
make_obs now takes dicts or strings

rmclaren committed:
make_obs now takes dicts or strings

nicholasesposito committed:
changes for range (#64)

Co-authored-by: Nicholas Esposito <esposito@hercules-login-4.hpc.msstate.edu>

rmclaren committed:
extending obs builder to handle rarer cases

rmclaren committed:
updated args for __main__ function

rmclaren committed:
Merge branch 'feature/obs_builder' of https://github.com/NOAA-EMC/bufr-query into feature/obs_builder

givelberg committed:
fizes several bugs, mostly related to config (#65)

Co-authored-by: Edward Givelberg <edwardg@hercules-login-2.hpc.msstate.edu>

rmclaren committed:
updated docs

rmclaren committed:
updated docs

rmclaren committed:
merged latest from feature/obsbuilder

rmclaren committed:
add config parameter to generated functions

rmclaren committed:
added obsbuilder test

rmclaren committed:
merged fetature/obsbuilder

rmclaren committed:
updates. Uses signature from obsbuilder override for external functions.

rmclaren committed:
removed create_obs_file_from_config and fixed the main function.

rmclaren committed:
Enhancements to ObsBuilder add_main_functions (#66)

* extending obs builder to handle rarer cases

* add config parameter to generated functions

* updates. Uses signature from obsbuilder override for external functions.

* removed create_obs_file_from_config and fixed the main function.

rmclaren committed:
updated create_obs_group

rmclaren committed:
updated create_obs_group

rmclaren committed:
actually use the cache_categories parameter in create_obs_group

rmclaren committed:
fixed some minor things

rmclaren committed:
updated documentation

rmclaren committed:
updated documentation

rmclaren committed:
updated documentation

rmclaren committed:
updated documentation

rmclaren committed:
updated documentation

rmclaren committed:
updated documentation

rmclaren committed:
updated documentation

rmclaren committed:
load config file .json before using it.

rmclaren committed:
load config file .json before using it.

rmclaren committed:
load config file .json before using it.

rmclaren committed:
merged develop

emilyhcliu committed:
Add Scan angle twist for CrIS and Code restructure.

emilyhcliu committed:
Merge branch 'develop' into feature/obs_builder_obs_group_cris

nicholasesposito reviewed (approved):
Looks good and makes sense.

PraveenKumar-NOAA reviewed (approved):
Looks good to me.

rmclaren commented on core/src/bufr/BufrReader/Exports/Variables/SensorScanAngleVariable.cpp:
```diff
@@ -16,13 +17,80 @@ namespace
+    // Radians to degree conversion
+    constexpr float radToDeg(float radians) {
+        return radians * (180.0f / M_PI);
+    }
```
Why constexpr? You are not using them that way...

emilyhcliu replied:
Fixed! Change `constexpr` to `static`

rmclaren commented on core/src/bufr/BufrReader/Exports/Variables/SensorScanAngleVariable.cpp:
```diff
@@ -16,13 +17,80 @@ namespace
+        4.77057f, 3.98517f, 3.19977f,
+        5.55597f, 0.0f,     2.41437f,
+        0.05818f, 0.84358f, 1.62897f
+    }; // unit is radians
```
naming convention for static const should be **FovDist**, **FovAng**.

emilyhcliu replied:
Fixed!

rmclaren commented on core/src/bufr/BufrReader/Exports/Variables/SensorScanAngleVariable.cpp:
```diff
@@ -16,13 +17,80 @@ namespace
+    void computeIasiScanAngles(std::vector<float>& scanang, const std::vector<int>& fovn,
+                               const std::vector<int>& scanpos, float start, float step, float stepAdj)
+    {
+        float tmp = -stepAdj;
+        for (size_t idx = 0; idx < fovn.size(); ++idx)
+       	{
+            if (scanpos[idx] % 2 == 1) tmp = stepAdj;
+            scanang[idx] = start + static_cast<float>((fovn[idx] - 1) / 4) * step + tmp;
+        }
```
Computation of **tmp** is not correct. It will only be -stepAdj until it enters the **if** statement the very first time. It will be positive forever afterword.

emilyhcliu replied:
True!!  Fixed.

rmclaren commented on core/src/bufr/BufrReader/Exports/Variables/SensorScanAngleVariable.cpp:
```diff
@@ -16,13 +17,80 @@ namespace
+ 
+       for (size_t idx = 0; idx < fovn.size(); ++idx)
+       {
+           const int for_idx = forn[idx] - 1;    // 0-based index
+           const int fov_idx = fovn[idx] - 1;    // 0-based index
+
+           // Degrees → Radians
+           float angle_offset_deg = static_cast<float>(for_idx) * step;
+           float scan_angle_rad = degToRad(start + angle_offset_deg);
+           float twist_angle_rad = degToRad(fov_ang[fov_idx] - angle_offset_deg);
```
Using python naming conventions. c++ convention is camelCase. So forIdx, fovIdx, angleOffsetDeg, scanAngleRad, etc...

emilyhcliu replied:
Fixed!

rmclaren commented on core/src/bufr/BufrReader/Exports/Variables/SensorScanAngleVariable.cpp:
```diff
@@ -39,89 +107,58 @@ namespace bufr {
-        // Get field-of-view number
-        std::vector<int> fovn(fovnObj->size(), DataObject<int>::missingValue());
-        for (size_t idx = 0; idx < fovnObj->size(); idx++)
+        if (sensor == "cris")
```
There are a lot of if statements like this in this function.. Would if make sense to have seperate functions, one for CRIS and one for IASI one for Generic and then call them from this function (get rid of the if statements)?

emilyhcliu replied:
OK.  Good suggestion.  I will try that!

emilyhcliu replied:
@rmclaren I modified the code to use fewer' if statements`. Please let me know if I can improve the code in any way. Thanks.

emilyhcliu replied:
Tagging @rmclaren I modified the code based on your suggestions.  It is ready for you to review again.

emilyhcliu committed:
Modify to reduce if statment

emilyhcliu committed:
fix bug found in CrIS scan angle calculation.

rmclaren commented on core/src/bufr/BufrReader/Exports/Variables/SensorScanAngleVariable.h:
```diff
@@ -12,14 +13,25 @@
+    using ComputeFn = std::function<void(
+        std::vector<float>&,               // Output: scan angles
+        const std::vector<int>&,           // Input: field-of-view numbers
+        const std::vector<int>*,           // Optional: field-of-regard numbers (nullptr if unused)
```
The accepted "modern" way of handling this in C++ is to use a `std::optional` (instead of using pointers). So this line of code would be something like `std::optional<const std::vector<int>>`. Or just default it with an empty vector.

Please see the following comments before changing this, however, as this advice may be overcome by events.

rmclaren replied:
ComputFn could probably be moved to the cpp file as you don't need it in the `public` interface. But its fine :)

rmclaren commented on core/src/bufr/BufrReader/Exports/Variables/SensorScanAngleVariable.h:
```diff
@@ -36,5 +48,23 @@ namespace bufr {
+        static void computeGeneric(std::vector<float>& scanang,
+                                   const std::vector<int>& fovn,
+                                   const std::vector<int>* forn,
+                                   float start, float step, float stepAdj);
```
Maybe these functions should just take the `const eckit::LocalConfiguration& conf` parameter?

rmclaren commented on core/src/bufr/BufrReader/Exports/Variables/SensorScanAngleVariable.cpp:
```diff
@@ -35,103 +61,62 @@ namespace bufr {
+        if (sensor == "cris" && conf_.has(ConfKeys::FieldOfRegardNumber)) {
+            auto& fornObj = map.at(getExportKey("fieldOfRegardNumber"));
+            fornVec = std::dynamic_pointer_cast<DataObject<int>>(fornObj)->getRawData();
+            fornPtr = &fornVec;
```
Each function could take the `conf` object and then do these steps (the ones necessary for the function). You would be **encapsulating** the knowledge of each data type inside their respective functions.

rmclaren commented on core/src/bufr/BufrReader/Exports/Variables/SensorScanAngleVariable.cpp:
```diff
@@ -35,103 +61,62 @@ namespace bufr {
+        if (it != dispatch.end()) {
+            it->second(scanang, fovn, fornPtr, start, step, stepAdj);
+        } else {
+            computeGeneric(scanang, fovn, nullptr, start, step, stepAdj);
```
Nice :)

emilyhcliu commented:
@rmclaren, You advised me to use the configuration for the function calculation for scan angles. 
Please elaborate a bit more on your advice.  Thanks....!

rmclaren commented:
@emilyhcliu Your functions for `computeIasi`, `computeCris` and `computeGeneric` would look like this:

```cxx
/// \brief Compute scan angle for IASI instrument
static void computeIasi(const eckit::LocalConfiguration& conf);

/// \brief Compute scan angle for CrIS instrument
static void computeCris(const eckit::LocalConfiguration& conf);

/// \brief Compute scan angle using generic (default) logic
static void computeGeneric(const eckit::LocalConfiguration& conf);
```

The each one would read what they need out of the conofiguration:

```cxx
static void computeIasi(const eckit::LocalConfiguration& conf)
{
  const float start = conf.getFloat(ConfKeys::ScanStart);
  const float step  = conf.getFloat(ConfKeys::ScanStep);
  // etc ....
```

emilyhcliu commented:
> @emilyhcliu Your functions for `computeIasi`, `computeCris` and `computeGeneric` would look like this:
> 
> ```c++
> /// \brief Compute scan angle for IASI instrument
> static void computeIasi(const eckit::LocalConfiguration& conf);
> 
> /// \brief Compute scan angle for CrIS instrument
> static void computeCris(const eckit::LocalConfiguration& conf);
> 
> /// \brief Compute scan angle using generic (default) logic
> static void computeGeneric(const eckit::LocalConfiguration& conf);
> ```
> 
> The each one would read what they need out of the conofiguration:
> 
> ```c++
> static void computeIasi(const eckit::LocalConfiguration& conf)
> {
>   const float start = conf.getFloat(ConfKeys::ScanStart);
>   const float step  = conf.getFloat(ConfKeys::ScanStep);
>   // etc ....
> ```

@rmclaren Thanks for the tips.  I will do so accordingly.  Thanks!

Commit:
Use configuration as input to computeFn function

emilyhcliu commented:
@rmclaren I have modified the code as you suggested
However, I have to pass both the `configuration` and the `map` into the function.
```
/// \brief Compute scan angle for IASI instrument
static void computeIasi(const eckit::LocalConfiguration& conf, const BufrDataMap& map);

/// \brief Compute scan angle for CrIS instrument
static void computeCris(const eckit::LocalConfiguration& conf, const BufrDataMap& map);

/// \brief Compute scan angle using generic (default) logic
static void computeGeneric(const eckit::LocalConfiguration& conf, const BufrDataMap& map)**
```

Notes:
I need the map so I can extract the data needed for the functions. So, I passed the `map` to the function.  But `exportKeys` are not accessible from the function, so I added the confKey information to the configuration pass it into the function.
